Dl3 reader for LST data

mstrzys opened the pull request:
DL3 reader; first for LST data only; to add a DL3 reader also for MAGIC data the filetype check needs a rework, which will go into a separate PR.

Commit:
feat: added DL3 data reader for LST dl3 files

- added an event class for LST dl3 files
- tested that class can read LST dl3 files
- entire chain not tested
- DL3 format for other instruments (e.g. MAGIC) can be slightly
  different and mya require a different reader

ToDo:
- MAGIC DL3 reader
- test entire chain with LST DL3 files

Commit:
- just made the LstDl3event class usable by the higher classes

Commit:
bug fix: tested the DL3 reader with LST-1 dl3 files

Found and corrected the following issues, some concern also previous
commits to the main branch:

- issue of the unti of the event time variable
- wrong unit of the ra event variable
- dl2 file reader had an error if the ra variable of the pointing
  position was not precomputed
- wrong overwriting of the setter function for the bkgmap maker in the
  bkg classes

Commit:
bug fix: wrong attribute in the stacked maps class: get_maps() attempted
to set attribute bkg_maps instead of the correct _bkg_maps

Commit:
- just remove unnecessary comma

IevgenVovk commented on src/pybkgmodel/camera.py:
Don't see why this file needed to be modified in this MR.

mstrzys replied:
Indeed, this modification is not necessary needed in this MR. Shall I remove it?

IevgenVovk commented on src/pybkgmodel/processing.py:
Don't see why this file needed to be modified in this MR.

mstrzys replied:
Same as above. Strictly speaking not needed. Shall I remove it or keep it?

IevgenVovk replied:
Exceptionally with @mstrzys we've agreed to leave it here. However, we've agreed to refactor those changes like

```python
@BkgMakerBase.bkg_map_maker.setter
    def bkg_map_maker(self, maker):
        if not isinstance(maker, WobbleMap):
            raise TypeError(f"Maker must be of type {WobbleMap}")
        BkgMakerBase.bkg_map_maker.fset(self, maker)
```
->
```python
@property
def bkg_map_maker(self):
    return self.__bkg_map_maker

@bkg_map_maker.setter
def bkg_map_maker(self, maker):
    if not isinstance(maker, WobbleMap):
        raise TypeError(f"Maker must be of type {WobbleMap}")
    self.__bkg_map_maker = maker
```

IevgenVovk replied:
There was a mistake in the example above; I see you have correctly used the protected instead of private variable for `bkg_map_maker` property. Very well.

IevgenVovk commented:
In general PR looks okey save the inserted docstring placeholders (useless unless replaced with actual docs). I would encourage in the future not to mix topical PRs like this with bug and style fixes in the unrelated classes and modules.

maxnoe commented on src/pybkgmodel/data.py:
```diff
@@ -557,13 +559,13 @@ def __init__(self, file_name, cuts=None):
     def get_obs_id(cls, file_name):
-        parsed = re.findall('.*dl3_LST-1.Run(\d+).fits', file_name)
+        # This code really sucks. need to make this more general...
+        parsed = re.findall('.*dl3_LST-1.Run(\d+).fits.gz', file_name)
```
Also `.` means "any character" in regexes, you need to escape if you really want `\.`.

LukasNickel replied:
Sorry, I accidentally pushed to this branch.

Commit:
<feat>: implement the dl3 reader

 - Modified the reader so it can read DL3 file regardless the instrument.
 - Data units are read from hdu header as long as provided.
 - Tested with LST and MAGIC DL3 files.

TODO:
 - Observatory location not part of the mandatory header keywords. May
   need to be provided in input file by the user. Will be part of future
pull request
 - Drift method implemented but not tested due to lack of corresponding data.

Commit:
<feat> Minor modifications to the reader for improved compatibility and
performance
 - added some functionalities suggested by Lukas Nickel, but adapted
   them to the current version
 - changed the function for retrieving the Obs ID from filename to
   reading it from the hdu header
 - changed the is_compatible, so it also identifies compressed files
 - added the erfaAstromInterpolator for the coordinate transform for
   speed up
 - added docstrings to all functions related to the DL3 reader

Commit:
<feat> adapted the model.py file to use the new DL3 reader.

Commit:
<feat> Added the Obs mode WOBBLE in addition to POINTING. Seems the Obs
modes are not to well defined e.g. the H.E.S.S. public data use WOBBLE
as the obs mode. Hence, added it to support such files until the allowed
modes are better defined.

mstrzys commented:
The latest changes allow the DL3 reader to not only read the LST files, but DL3 files in general. I tested it with LST, H.E.S.S. (public data release) and MAGIC DL3 files. 

Furthermore added the suggestions from Lukas regarding the interpolator for the coordinate transform and the changes regarding the Obs ID retrieval; changed the compatibility check to not reject compressed files as well.

mstrzys commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +534,178 @@ def load_events(cls, file_name, cuts):
+                        event_data[name] = evt_data[key].to_numpy()
+
+                # Event times need to be converted from LST Epoch
+                LST_EPOCH = astropy.time.Time('2018-10-01T00:00:00', scale='utc')
```
Outdated: LST_EPOCH does not exists any more. Now called: "ref_epoch".

mstrzys commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +534,178 @@ def load_events(cls, file_name, cuts):
+                # Event times need to be converted from LST Epoch
+                LST_EPOCH = astropy.time.Time('2018-10-01T00:00:00', scale='utc')
+                event_data['mjd'] = astropy.time.Time(event_data['mjd'], format='unix')
+                event_data['mjd'] = astropy.time.Time((event_data['mjd'].unix + LST_EPOCH.unix),
```
Right that was a bit lazy solution as I read the data from the dataframe into the dictionary and directly gave this dictionary the final name. I agree it is confusing and change the way the event times are read.

IevgenVovk replied:
OK, this looks resolved

mstrzys commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +534,178 @@ def load_events(cls, file_name, cuts):
+                                                      ).mjd
+
+                # Compute the telescope pointing positions for each event
+                lst_time = astropy.time.Time(event_data['mjd'], format='mjd')
```
This comment is outdate. This line is not part of the latest version of this merge request.

Commit:
refactor: change variables and data handling to accommodate comments
from the corresponding PR (#7).

- change name from Dl2 to DL2 to agree with the DL3 naming
- changed to way the event times are read from DL3 files to be
  consistent with the 'mjd' naming of the dictionary
- the functionality of the code was not changed

HealthyPear commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +537,236 @@ def load_events(cls, file_name, cuts):
+        """
+
+        ext = Path(file_name).suffixes
+        compatible = ".fits" in ext
```
I don't think this is a robust way to check for compatibility, perhaps `astropy.fits` machinery is better?
https://docs.astropy.org/en/stable/io/fits/usage/verification.html

mstrzys replied:
In general I agree that it is not the most robust way. However, any deeper inspection of the files will require us to first define a strategy down to which level we like to check these files. While supporting the idea, maybe it going beyond the scope of this PR. What do you think? Maybe open a separate issue? Do you know whether gammapy already has a validation which we could use?

HealthyPear replied:
For what concerns this specific line what I meant is that you are already importing here the `astropy.fits` module so doing something like

```python
try:
    with fits.open(input_file_path) as file:
        pass
    compatible = True
except OSError:
    compatible = False
```

should  be more reliable than just checking for the file extension (I didn't test it, but it might be that if the extension is not ".fits" a file can still a valid FITS file).

Technically `astropy` should already check some basic things (like e.g. the presence of a primary HDU).

For what concerns _gammapy_, if the FITS file in input is supposed to follow GADF I think it already checks for that schema, but I would not put too much in this PR.

mstrzys replied:
Adapted the solution and added it to the PR.

IevgenVovk replied:
Looks OK now.

maxnoe commented:
Hi,

I got a notification here and had a cursory glance, but what I saw made me very nervous about timing information.

You add and subtract raw mjd and unix values at different points in the code here instead of using astropy Time and TimeDelta objects.

This will result in timestamps that are off by the number of leap seconds between those two times. Never do that.

maxnoe commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +537,236 @@ def load_events(cls, file_name, cuts):
+                            event_data[name] *= u.one
+
+                # Event times need to be converted from Instrument reference epoch
+                ref_epoch = astropy.time.Time(evt_head['MJDREFI']+evt_head['MJDREFF'], format='mjd')
```
Use the two argument form of `astropy.time.Time`, otherwise you will have precision issues:

```
ref_epoch = astropy.time.Time(evt_head['MJDREFI'], evt_head['MJDREFF'], format='mjd')
```

maxnoe replied:
also, you need to read the time scale

mstrzys replied:
added, though I found that in the header of the DL3 files the TIMESYS is given in all-caps whereas astropy prefers lower cases, is there a reason for why it is all-caps in the LST files?

maxnoe replied:
Because that's what the FITS standard is prescribing

maxnoe commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +537,236 @@ def load_events(cls, file_name, cuts):
+                # Event times need to be converted from Instrument reference epoch
+                ref_epoch = astropy.time.Time(evt_head['MJDREFI']+evt_head['MJDREFF'], format='mjd')
+
+                event_data['mjd'] = astropy.time.Time((evt_data['TIME'].to_numpy() 
```
Do not do this! this results in times that are off by the number of leap seconds between the reference epoch and the event time.

The correct solution is `(ref_epoch + evt_data["TIME"].quantity).mjd`

mstrzys replied:
fixed

maxnoe commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +537,236 @@ def load_events(cls, file_name, cuts):
+                                                      ).mjd * u.d
+
+                # Adapt
+                evt_time = astropy.time.Time(event_data['mjd'], format='mjd')
```
now you create a time object a second time.

Just to `evt_time = ref_epoch + evt_data["TIME"].quantity` and then `evt_data["mjd"] = evt_time.mjd`

maxnoe replied:
All of this is also already implemented in Gammapy, e.g. here: https://github.com/gammapy/gammapy/blob/main/gammapy/utils/time.py

mstrzys replied:
Simplified according to your suggestion.

maxnoe commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +537,236 @@ def load_events(cls, file_name, cuts):
+            try:
+                evt_hdu = input_file["EVENTS"]
+                evt_head = evt_hdu.header
+                evt_data = pandas.DataFrame(evt_hdu.data)
```
use an astropy table to have support for the quantities and metadata stored in the hdu

mstrzys replied:
Thanks for that hint, simplified it a lot. Adapted.

mstrzys committed:
FIX:

 - changed the time handling following Max' recommendations in the discussion of the PR
 - changed the way the fits file compatibility is checked following Michele's recommendations in the discussion of the PR
 - changed the setter for the bkg_map_makers following Ievgen's recommendation

TODO:
 - Test with data

maxnoe commented on src/pybkgmodel/data.py:
```diff
@@ -654,25 +662,20 @@ def load_events(cls, file_name):
-                                                       + ref_epoch.unix),
-                                                      scale='utc',
-                                                      format='unix'
+                event_data['mjd'] = astropy.time.Time((evt_data['TIME'].quantity 
```
The outer time here is unnecessary.

The result of `ref_epoch + evt_data["TIME"].quantity` is a time, you are just making a copy again.

mstrzys replied:
Removed

maxnoe commented on src/pybkgmodel/data.py:
```diff
@@ -654,25 +662,20 @@ def load_events(cls, file_name):
                 if evt_head['OBS_MODE'] in ('POINTING', 'WOBBLE'):
 
-                    alt_az_frame = AltAz(obstime=evt_time,
+                    alt_az_frame = AltAz(obstime=event_data['mjd'],
```
I don't think you can pass a raw mjd value to `AltAz(obstime=)`, this needs to be a `Time` instance.

mstrzys replied:
Indeed, corrected.

mstrzys committed:
FIX:

  - simplified the reading of the data from the fits files with astropy.Tables follwing Max' suggestion
  - fixed the obstime object in the AltAz frame definition following Max' comment in the discussion of the PR

mstrzys committed:
FIX:

 - fixed the setter function of the bkg_map_maker

mstrzys committed:
FIX:

 - debugging the setter for the bkg_map_maker
 - current version not running

mstrzys committed:
FIX:

 - intermediate commit for checking the updated setter

mstrzys committed:
FIX:

 - ready for further review
 - setter for bkg_map_maker fixed
 - time scale added to ref_epoch
 - checked that the code runs with with LST-1 data

maxnoe commented on src/pybkgmodel/camera.py:
```diff
@@ -375,7 +375,7 @@ def get_pixel_coords(self):
         x = (self.xedges[1:] + self.xedges[:-1]) / 2
         y = (self.yedges[1:] + self.yedges[:-1]) / 2
-        xx, yy = np.meshgrid(x, y, indexing='ij')
+        xx, yy = numpy.meshgrid(x, y, indexing='ij')
```
I think this is the wrong resolution of a merge conflict? 

The import at the top is `import numpy as np`

mstrzys replied:
As entire merge went wrong. I reset the merge. New version should not have this mistake.

maxnoe commented on src/pybkgmodel/data.py:
```diff
@@ -312,11 +314,11 @@ def load_events(cls, file_name, cuts):
                     # Post processing
-                    event_data['true_zd'] = np.degrees(event_data['true_zd'])
-                    event_data['true_az'] = np.degrees(event_data['true_az'])
+                    event_data['true_zd'] = numpy.degrees(event_data['true_zd'])
```
same as above

mstrzys replied:
As entire merge went wrong. I reset the merge. New version should not have this mistake.

maxnoe commented:
@mstrzys I think something went seriously wrong in your last merge commit, all the changes related to reading LST DL3 data were removed

Commit:
Merge branch 'main' into dl3_reader

Commit:
FIX:

  - changed the remaining numpy's to np, not covered by the merger

Commit:
FIX:

 - removed the np.bool for bool inherited from the main branch
 - tested with LST1 data and ensure that it runs properly

maxnoe commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +543,235 @@ def load_events(cls, file_name, cuts):
+                                              )
+
+                evt_time = evt_data['TIME'].quantity + ref_epoch
+                event_data['mjd'] = evt_time.mjd * u.d
```
I am not sure, but I think this mjd will now depend on the timescale of the input file, and I guess you use this value internally and assume it's in UTC or some other fixed timescale?

Why are you storing the mjd in any case and not the full time object?

mstrzys replied:
Issue with time scale is not addressed and it now ensures that the event_data['mjd'] are in utc regardless the input scale.

We decided to not store the full time object for consistency within the event_data dictionary and for an -admittedly minor - performance boost.

maxnoe commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +543,235 @@ def load_events(cls, file_name, cuts):
+                                                            * len(event_data['pointing_zd'])
+                                                            ) * u.deg
+
+                elif evt_head['OBS_MODE'] == 'DRIFT':
```
Just raise an error to let users know this is not supported instead of providing preliminary or invalid data (same for the next else block, this should be an actual error, not a print statement)

mstrzys replied:
Drift mode is now tested, so no need for an error there. I however changed it for the case the file contains an unsupported mode. This will now raise an error instead of giving a warning.

Commit:
<fix>
- addressed the comments in the MR by Max regarding the time scale read from the file and ensured it is converted to utc
- tested the drift scan mode using a test simulation
- using a not implemented methods will now raise an error instead of jsut printing a warning
- Ievgen and Marcel decided to not keep the time object (comment by Max), but to keep the current implementation to ensure the consistency of the variables and for a performance boost

IevgenVovk commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +543,230 @@ def load_events(cls, file_name, cuts):
+    Parameters
+    ----------
+    file_name: str
+            Name of the DL3 file to use.
```
Indent is too large

mstrzys replied:
fixed

IevgenVovk replied:
Looks like the corresponding change is not committed yet.

IevgenVovk commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +543,230 @@ def load_events(cls, file_name, cuts):
+            with fits.open(ext) as file:
+                pass
+            compatible = True
+        except OSError:
```
This code will break if it's not `OSError` (`compatible` won't be defined). Suggest catching this edge case too.

mstrzys replied:
Now set the compatibility to False by default catching this case.

IevgenVovk replied:
Like this the `OSError` catching is not needed... OK

IevgenVovk commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +543,230 @@ def load_events(cls, file_name, cuts):
+        }
+
+        with fits.open(file_name, memmap=False) as input_file, \
+            erfa_astrom.set(ErfaAstromInterpolator(1 * u.s)):
```
Will affect the rest of `pybkgmodel` code as you don't reset this back. Suggest nesting the relevant parts into a context manager (see [astropy docs](https://docs.astropy.org/en/latest/coordinates/index.html#improving-performance-for-arrays-of-obstime)).

mstrzys replied:
I think the interpolation is already encapsulated in a context manager. Please mind that the previous line ends with a "\", so l. 648 just continues the previous line and thus should be part of the context manager.

IevgenVovk replied:
I see. This is not very readable, but OK

IevgenVovk commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +543,230 @@ def load_events(cls, file_name, cuts):
+
+
+                    event_data['pointing_ra'] = np.array(
+                                                            [evt_head['RA_PNT']] \
```
Please select one way to indent the code
This 
```
alt_az_frame = AltAz(obstime=evt_time,
                    location=obs_loc)
```
and this
```
event_data['pointing_dec'] = np.array(
                                        [evt_head['DEC_PNT']] \
                                        * len(event_data['pointing_zd'])
                                        ) * u.deg
```
are clearly different.

mstrzys replied:
Updated the indentation and the alignment.

IevgenVovk commented on src/pybkgmodel/data.py:
```diff
@@ -501,23 +543,230 @@ def load_events(cls, file_name, cuts):
+                      "Empty arrays will be returned.")
+
+        finite = [np.isfinite(item) for key, item  in event_data.items() if item is not None]
+        all_finite = np.prod(finite, axis=0, dtype=bool)
```
Maybe a simple `all(finite)` would do?

mstrzys replied:
Well, this was carried over from a function you implemented. I am now not sure about this case, but if you mean np.prod should be replaced by np.all, then these function are not the same in all cases.

IevgenVovk replied:
As elements of `finite` are all bool, result of `np.prod(finite)` is the same as `np.all(finite)` - with the latter better reflecting the logic of this check. Aside of this, current version should be workable.

Commit:
<fix> addressin some comments in the MR
- changed formatting
- prevented crash in case the file format is not correct but does not trigger the OSError

Commit:
<style> updated the style of the changes to address commments in the MR
